quarkchain/evm/specials.py

```diff
@@ -237,16 +237,32 @@ def proc_current_mnt_id(ext, msg):
 def proc_transfer_mnt(ext, msg):
     from quarkchain.evm.messages import apply_msg
 
-    # Data must be >= 96 bytes
-    if msg.data.size < 96:
-        return 0, 0, []
-    gascost = 3
-    if msg.gas < gascost:
+    # Data must be >= 96 bytes and static call not allowed
+    if msg.data.size < 96 or msg.static:
         return 0, 0, []
     to = utils.int_to_addr(msg.data.extract32(0))
     mnt = msg.data.extract32(32)
     value = msg.data.extract32(64)
     data = msg.data.extract_all(96)
+
+    num_zero_bytes = data.count(ascii_chr(0))
+    num_non_zero_bytes = len(data) - num_zero_bytes
+    # Intrinsic gas
+    gascost = (
+        opcodes.GTXCOST
+        + opcodes.GTXDATAZERO * num_zero_bytes
+        + opcodes.GTXDATANONZERO * num_non_zero_bytes
+    )
+    if value > 0:
+        gascost += opcodes.GCALLVALUETRANSFER
+        if not ext.account_exists(to):
+            gascost += opcodes.GCALLNEWACCOUNT
+    # Out of gas
+    if msg.gas < gascost:
+        return 0, 0, []
+    # Handle insufficient balance or exceeding max call depth
+    if ext.get_balance(msg.sender, token_id=mnt) < value or msg.depth >= vm.MAX_DEPTH:
```

Comment:
I think we may still add GSTIPENT in submsg gas (assuming GCALLVALUETRANSFER > GSTIPENT) since GSTIPENT may not be supplied when calling the contract.

Further, we may not allow "to" to be any special contracts to eliminate the reentrancy issue (or other possible corner cases).

Reply:
users can set `to` as another address which in turn calls a special contract. we control this kind of recursive invocation

Reply:
1. sounds good
1. seems enough to limit `to` not to be itself, otherwise if we limit `to` not be **any** special address, we have to deal with potential conflicts in the future if we add more special addresses.